src/utils/songInfo.ts

```diff
@@ -15,3 +15,102 @@ export const getPlaybackNotes = (
     .filter(track => track.channel === playbackChannel)
     .flatMap(track => track.notes);
 };
+
+type SmartMapping = number[];
+export type SmartMappingChangeEvent = {
+  time: number;
+  mapping: SmartMapping;
+};
+
+// TODO
+export const getSmartNotes = (
+  notes: Note[],
+  changeEvents: SmartMappingChangeEvent[]
+): SmartNote[] => {
+  const smartNotes: SmartNote[] = notes.map(note => {
+    return Object.assign({}, note, {
+      ...notes,
+      smartKey: Math.floor(Math.random() * Math.floor(7)), // TODO: find the corresponding key
+    });
+  });
+
+  return smartNotes;
+};
+
+// TODO
+export const getSmartMappingChangeEvents = (
```

Comment:
This method returns an array of change events
- Ideally, the change should take place slightly before the actual note play time (in case there is slight delay in the change)